sql/core/src/test/scala/org/apache/spark/sql/DataFrameWriterV2Suite.scala

```diff
@@ -839,4 +839,18 @@ class DataFrameWriterV2Suite extends QueryTest with SharedSparkSession with Befo
       condition = "CALL_ON_STREAMING_DATASET_UNSUPPORTED",
       parameters = Map("methodName" -> "`writeTo`"))
   }
+
+  test("SPARK-49975: write should fail with legacy store assignment policy") {
+    withSQLConf((SQLConf.STORE_ASSIGNMENT_POLICY.key, "LEGACY")) {
+      checkError(
+        exception = intercept[AnalysisException] {
+          spark.range(10)
+            .writeTo("table_name")
```

Comment:
what was the behavior before? Do we add hard cast for the input query?

Reply:
This test passes on master branch. I was wrong about what's allowed with LEGACY store assignment policy.


The following is allowed with either SQL or DataFrameWriteV2 API when `source` has the same schema as `v2Table`.

```sql
insert into v2Table select * from source;
```

```scala
val df = spark.table("source")
df.writeTo("v2Table").append()
```

Reply:
OK so the current logic is: if Spark needs to adjust the input query before writing to the table (type coercion), then we disallow LEGACY store assignment. Otherwise, we don't check store assignment mode at all. This is consistent across multiple APIs and seems fine?